src/transformStateWithClones.js

```diff
@@ -5,9 +5,29 @@
  * @param {Object[]} actions
  *
  * @return {Object[]}
+ *
  */
 function transformStateWithClones(state, actions) {
   // write code here
+  const stateHistory = [];
+  const copyState = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(copyState, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      for (const element of action.keysToRemove) {
+        delete copyState[element];
+      }
+    } else if (action.type === 'clear') {
+      for (const key of Object.keys(copyState)) {
+        delete copyState[key];
+      }
+    }
+    stateHistory.push(Object.assign({}, copyState));
```

Comment:
It's better to perform the push to `stateHistory` only once after the current action is processed, outside of the `switch` block. This will adhere to the 'Don't Repeat Yourself' principle.